[RFC] remove vim_tolower/etc functions with broken locale-dependent behavior

bfredl opened the pull request:
ref https://github.com/neovim/neovim/issues/6474#issuecomment-292700689 and vim issue https://groups.google.com/forum/#!topic/vim_dev/sn0MDYfETCs . Note this is not the "standard" Turkish i issue where locale-dependent behavior cases surprises, rather these functions are broken and leaves i and I _unchanged_ no matter the value of `casemap`. The fix is simple is this codebase: delete the `vim_tolower` wrappers and just always use the multibyte functions which have the correct semantics.

In vim the global names are on the form `MB_TOLOWER()`, but as it is not macros anymore, use `mb_tolower()`

bfredl committed:
mbyte: replace vim_tolower with mb_tolower handling locale correctly

bfredl committed:
test: add tests for gu/gU behavior in Turkish locale

ZyX-I commented on test/functional/normal/lang_spec.lua:
```diff
@@ -0,0 +1,48 @@
+
+  describe('works in Turkish locale', function()
+    before_each(function()
+      execute('lang ctype tr_TR.UTF-8')
```
I would suggest to use `exc_exec` and checking for it returning zero, not for error message.

bfredl replied:
:+1: I will also update `viml/lang_spec.lua` which this was copied from.

justinmk commented:
I guess the test from 8.0.0552 can be included here.

bfredl commented:
yes, even though it only tests the `casemap=` case despite I said also `casemap=keepascii` is broken..

bfredl committed:
vim-patch:8.0.0243

Problem:    When making a character lower case with tolower() changes the byte
            cound, it is not made lower case.
Solution:   Add strlow_save(). (Dominique Pelle, closes vim/vim#1406)

https://github.com/vim/vim/commit/cc5b22b3bfdc0e9e835cf7871166badda31447bd

Join almost identical strup_save and strlow_save functions to one
Function.

bfredl committed:
vim-patch:8.0.0552

Problem:    Toupper and tolower don't work properly for Turkish when 'casemap'
            is empty. (Bjorn Linse)
Solution:   Check the 'casemap' options when deciding how to upper/lower case.

https://github.com/vim/vim/commit/3317d5ebbe8304da82b8088446060afcae0012af

vim-patch:8.0.0553

Problem:    Toupper/tolower test with Turkish locale fails on Mac.
Solution:   Skip the test on Mac.

https://github.com/vim/vim/commit/9f4de1f5435b900e43e19766da1a5bed4686cf44

vim-patch:8.0.0554

Problem:    Toupper and tolower don't work properly for Turkish when 'casemap'
            contains "keepascii". (Bjorn Linse)
Solution:   When 'casemap' contains "keepascii" use ASCII toupper/tolower.

https://github.com/vim/vim/commit/1cc482069a3407132aeb43a55d6dc284153e79c7

vim-patch:8.0.0555

Problem:    Toupper/tolower test fails on OSX without Darwin.
Solution:   Skip that part of the test also for OSX. (Kazunobu Kuriyama)

https://github.com/vim/vim/commit/d2381a2cadb9ef359ad5efb916734c635b29bd13

bfredl committed:
lint: fix clint errors around mb_tolower calls

bfredl commented:
Hmm, it fails on mac (always ASCII behavior), should I skip the test just like vim does?

bfredl commented:
Now all the tests pass.
